### PR TITLE
Implemented JSON API (Valerii Riabokon)

### PR DESCRIPTION
**Pull request by starkaexname**
**Valerii Riabokon**

-to challenge_api added:
_Classes: Vehicle
Functions: vehicle_features_post, clean_input_data, is_valid, saving_vehicles, create_app -added test_routes: tests for challenge_api_
-added folder 'database' for JSON files
-updated requirements.txt

# Description
Elapsed time: 5h

### CODE REVIEW

File: src/core/challenge_api.py [hunk 1]
```diff
@@ -1,23 +1,108 @@
...
+except:
+    from src.utils.api_decorators import ApiDecorators
 
 api = Blueprint("challenge_api", __name__)
 
 
+def create_app(config="dev"):
+    app = Flask(__name__)
+    app.register_blueprint(api)
+    return app
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:19]: The flask app is already created in this project and the `challenge_api` is supposed to just implement this very api.
Have you read the `README.md`?

File: src/core/challenge_api.py [hunk 2]
```diff
@@ -1,23 +1,108 @@
...
+import os
+from flask import Blueprint, make_response, Flask
+from flask import request
+from json import *
+from jsonschema import Draft7Validator
+
+try:
+    from utils.api_decorators import ApiDecorators
+except:
+    from src.utils.api_decorators import ApiDecorators
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:11]: I would consider this as a bad practice. Run it like described in the `README.md` and you won't need to do such thing.

File: src/core/challenge_api.py [hunk 3]
```diff
@@ -1,23 +1,108 @@
...
+            print("Aborted")
+
+
 @api.route("/challenge", methods=["POST"])
 @ApiDecorators.require_customer_id
-def vehicle_features_post(user_id: str):
-    """
-    Please see the README.md
-    Also see `json/vehicle-features.v1.schema.json` and `json/vehicle-features.v1.example.json`
+def vehicle_features_post(user_id):
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:65]: You are not supposed to get rid of the type annotation. BTW the entire implementation is missing type annotation.

File: src/core/challenge_api.py [hunk 4]
```diff
@@ -1,23 +1,108 @@
...
+    app = Flask(__name__)
+    app.register_blueprint(api)
+    return app
+
+
+class Vehicle:
+    def __init__(self, vehicle_id, vehicle_features, user_id):
+        self.id = vehicle_id
+        self.features = vehicle_features
+        self.user_id = user_id
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:26]: The user id basically is not part of the Vehicle. I understand the desire to store it somewhere it is probably the wrong place.

File: src/core/challenge_api.py [hunk 5]
```diff
@@ -1,23 +1,108 @@
...
+        self.folder_path = f"database/{self.user_id}"
+        self.file_path = f"{self.folder_path}/{self.id}.json"
+
+    def __str__(self):
+        return self.id
+
+    def save(self):
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:33]: If we consider the `Vehicle` class as a straight data class, what would your approach be to implement the persisting functionality?

File: src/core/challenge_api.py [hunk 6]
```diff
@@ -1,23 +1,108 @@
...
+        json_data = dict(id=self.id, features=self.features)
+        if not os.path.isdir(self.folder_path):
+            os.mkdir(self.folder_path)
+        if not os.path.isfile(self.file_path):
+            out_file = open(f"{self.file_path}", "w")
+            json.dump(json_data, out_file)
+            out_file.close()
+
+    def get(self):
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:42]: ^

File: src/core/challenge_api.py [hunk 7]
```diff
@@ -1,23 +1,108 @@
...
+        try:
+            file_json = open(f"{self.file_path}", "r")
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:44]: self.file_path is already of type `str`

File: src/core/challenge_api.py [hunk 8]
```diff
@@ -1,23 +1,108 @@
...
+            data_json = json.load(file_json)
+            file_json.close()
+            return data_json
+        except FileNotFoundError:
+            print("Error: The file does not exist")
+
+    def delete(self):
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:51]: Unused as far as I can tell

File: src/core/challenge_api.py [hunk 9]
```diff
@@ -1,23 +1,108 @@
...
+
 @api.route("/challenge", methods=["POST"])
 @ApiDecorators.require_customer_id
-def vehicle_features_post(user_id: str):
-    """
-    Please see the README.md
-    Also see `json/vehicle-features.v1.schema.json` and `json/vehicle-features.v1.example.json`
+def vehicle_features_post(user_id):
+    vehicles = clean_input_data()  # format input data
+    if is_valid(vehicles):  # validate JSON schema
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:67]: Big yes for the validation 👍

File: src/core/challenge_api.py [hunk 10]
```diff
@@ -1,23 +1,108 @@
...
+            return make_response("500", 500)
+    else:
+        return make_response("400", 400)
+
+
+def clean_input_data():
+    input_data = dict()
+    try:
+        input_data = request.get_json()
+    except:
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:83]: You do not want to use catch all exception handlings in general.

File: src/core/challenge_api.py [hunk 11]
```diff
@@ -1,23 +1,108 @@
...
+        input_data = request.get_data(as_text=True)
+        input_data = input_data[input_data.find("{") : input_data.rfind("}") + 1]
+        input_data = json.loads(input_data)
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:86]: You thought of a fallback, that's good. We probably can never rely on correct input from others^^

File: src/core/challenge_api.py [hunk 12]
```diff
@@ -1,23 +1,108 @@
...
+    finally:
+        input_data["vehicles"] = input_data.pop("vehicle")
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:88]: I do not quite understand this.

File: src/core/challenge_api.py [hunk 13]
```diff
@@ -1,23 +1,108 @@
...
 
-    :param user_id: The id of the customer sending the request
-    :return: ???
-    """
+def is_valid(input_data):
+    with open("json/vehicle-features.v1.schema.json") as sc:
+        schema = load(sc)
+    validator = Draft7Validator(schema)
+    if validator.is_valid(input_data):
+        return True
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:97]: Good to see that you are validating the schema. I would prefer the `validator` to be a singleton. If your concern is that this would get a bottleneck then there probably still is a better way than reading the file and initializing the validator over and over again

File: src/core/challenge_api.py [hunk 14]
```diff
@@ -1,23 +1,108 @@
...
 
-    # TODO Implement the challenge
 
-    return make_response("OK", 200)
+def saving_vehicles(vehicles, user_id):
+    list1 = list()
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:101]: Please use meaningful names

File: src/core/challenge_api.py [hunk 15]
```diff
@@ -1,23 +1,108 @@
...
+    for vehicle in vehicles["vehicles"]:
+        if vehicle["id"] in list1:
+            raise FileExistsError
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:104]: I Like the thought but it is certainly no `FileExistsError`

File: src/core/challenge_api.py [hunk 16]
```diff
@@ -1,23 +1,108 @@
...
+        else:
+            list1.append(vehicle["id"])
+    for vehicle in vehicles["vehicles"]:
+        Vehicle(vehicle["id"], vehicle["features"], user_id).save()
```
Comment (markus-fuereder) [inline src/core/challenge_api.py:108]: I'm not sure if the validator does check the content as well, would loved to see some validation as well.

File: src/core/challenge_api.py (hunk 14 above)
Comment (markus-fuereder) [inline src/core/challenge_api.py:101]: In general there are better ways to check duplicates, but it is good that you thought of it.

File: tests/functional/test_routes.py [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+class FuncTest(TestCase):
+    def setUp(self):
+        self.app = create_app(config="test")
+        self.client = self.app.test_client()
+
+    def test_index(self):
+        with open("json/vehicle-features.v1.example.json") as data:
+            res = self.client.post("/challenge", data=data, content_type="application/json")
+            assert res.status_code == 200
+
```
Comment (markus-fuereder) [inline tests/functional/test_routes.py:15]: Good to see you doing tests. A single one would not do the trick, but for the challenge it is fair enough to show that you would do test.

**Review by markus-fuereder — CHANGES_REQUESTED** [review]
Part of the challenge is to know the task and to know what to do. You are supposed to read and understand the instructions in the `README.md`.

In general the implementation is missing type annotation.
A lot of areas are hard to follow because of not ideal naming choices and the lack of meaningful comments. If first would be the case I would not mind the latter, since I personally believe that well written code does not need an explanation.

You updated the `requirements.txt` and added a lot of unnecessary bloat. Besides that you decided to update some versions of the already specified packages. This is quit dangerous in a real world application and would have needed a reason to do it.

In summary I'm not very impressed by your solution.
Your current level of expertise is not what we are looking for, so I unfortunately have to tell you, that you did not pass the challenge and we do not consider working with you.

I took the time to actually review the submission and hope you can take away something from it. Thanks for your interest and your effort.

All the best.
